### PR TITLE
feat: implement ip-router-interface requirer side

### PR DESCRIPTION
**Pull request by patriciareinoso**
# Description

Implements the requirer side of the `ip-router-interface` relation.
UPF IP address and gateway are no longer config parameters. They will be obtained from relation.
Config parameters to specify the user plane network config were added.
We will inform the sdcore-router of the user plane network config using the `request_network` method in the `ip-router-interface` library.
We will get the UPF config from the access network entry on the routing table. 
We obtain the routing table using the `get_routing_table` method in the `ip-router-interface` library o the content in `RoutingTableUpdatedEvent`

[Please include a summary of the change. Please also include relevant motivation and context. List any dependencies that are required for this change.
](https://docs.google.com/document/d/1nJBEfPw-k9YDWpqV-qrhZnfsJXyrnrjS-Q2SNnhdPrY/edit)

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -33,8 +38,10 @@
 CONFIG_FILE_NAME = "gnb.conf"
 GNB_INTERFACE_NAME = "gnb"
 GNB_NETWORK_ATTACHMENT_DEFINITION_NAME = "gnb-net"
+IP_ROUTER_RELATION_NAME = "ip-router"
```
Comment (patriciareinoso) [inline src/charm.py:41]: To check

File: src/charm.py [hunk 2]
```diff
@@ -135,15 +154,26 @@ def _configure(self, event: EventBase) -> None:
...
+
+    def _update_upf_config(self, event: RoutingTableUpdatedEvent) -> None:
+        ip_router_relations = self.model.relations.get(IP_ROUTER_RELATION_NAME)
+        if not ip_router_relations:
+            logger.info("No %s relations found.", IP_ROUTER_RELATION_NAME)
+            return
+        routing_table = event.routing_table.get("networks", {})
+
+        if networks := routing_table.get(ACCESS_NETWORK_NAME, []):
+            network = networks[0]
```
Comment (patriciareinoso) [inline src/charm.py:173]: To check

File: src/charm.py [hunk 3]
```diff
@@ -197,6 +227,34 @@ def _network_attachment_definitions_from_config(self) -> list[NetworkAttachmentD
...
+
+    def _get_upf_ip_address_and_gateway_from_ip_router(self) -> Tuple[str, str]:
+        ip_router_relations = self.model.relations.get(IP_ROUTER_RELATION_NAME)
+        if not ip_router_relations:
+            logger.info("No %s relations found.", IP_ROUTER_RELATION_NAME)
+            return ("", "")
+        routing_table = self._router_requirer.get_routing_table()
+        if networks := routing_table.get(ACCESS_NETWORK_NAME, []):
+            network = networks[0]
+            return network["network"], network["gateway"]
```
Comment (patriciareinoso) [inline src/charm.py:255]: to check

File: src/charm.py [hunk 4]
```diff
@@ -368,10 +431,20 @@ def _get_invalid_configs(self) -> list[str]:  # noqa: C901
...
+            return
+        if not upf_ip_address:
+            logger.warning("UPF IP address is missing.")
+            return
+        if not upf_gateway:
+            logger.warning("UPF gateway is missing.")
+            return
         self._exec_command_in_workload(
-            command=f"ip route replace {self._get_upf_ip_address_from_config()} via {self._get_upf_gateway_from_config()}"  # noqa: E501
+            command=f"ip route replace {upf_ip_address} via {upf_gateway}"
```
Comment (patriciareinoso) [inline src/charm.py:447]: to check

File: src/charm.py [hunk 5]
```diff
@@ -135,15 +154,26 @@ def _configure(self, event: EventBase) -> None:
             usim_sequence_number=self._get_usim_sequence_number_from_config(),  # type: ignore[arg-type]  # noqa: E501
             sst=self._get_sst_from_config(),  # type: ignore[arg-type]
             tac=self._get_tac_from_config(),  # type: ignore[arg-type]
-            upf_gateway=self._get_upf_gateway_from_config(),  # type: ignore[arg-type]
-            upf_ip_address=self._get_upf_ip_address_from_config(),  # type: ignore[arg-type]
+            upf_gateway=upf_gateway,
+            upf_ip_address=upf_ip_address,
```
Comment (patriciareinoso) [inline src/charm.py:158]: To check: if the upf gatway and ip address don't exist these parameters will be empty in config file.

File: tests/unit/test_charm.py [hunk 6]
```diff
@@ -218,35 +233,182 @@ def test_given_default_config_when_config_changed_then_status_is_active(
...
-                "upf-ip-address": upf_ip_address,
-                "upf-gateway": upf_gateway,
+                "user-plane-network": user_plane_network,
+                "user-plane-gateway": user_plane_gateway,
             }
         )
+        networks = [{"network": user_plane_network, "gateway": user_plane_gateway}]
+        patch_request_network.assert_called_with(
+            requested_networks=networks, custom_network_name="ip-router"
+        )
```
Comment (patriciareinoso) [inline tests/unit/test_charm.py:346]: Add test using default values

File: src/charm.py [hunk 7]
```diff
@@ -197,6 +227,34 @@ def _network_attachment_definitions_from_config(self) -> list[NetworkAttachmentD
             ),
         ]
 
+    def _request_user_plane_network_to_ip_router(self) -> None:
```
Comment (patriciareinoso) [inline src/charm.py:230]: Question for @Kayra1  after sprint!!
Context:
on a config change we will request a new network and then get the routing table to get the UPF network information, to create the route and write config file in the container.
However, requesting a new network will trigger a new routing table updated event , in which we will get the routing table from the event and get the UPF network information to create the route and write config file in the container. AGAIN.

File: metadata.yaml [hunk 8]
```diff
@@ -33,6 +33,8 @@ provides:
 requires:
   fiveg-n2:
     interface: fiveg_n2
+  ip-router:
+    interface: ip_router_user_plane
```
Comment (patriciareinoso) [inline metadata.yaml:37]: to check

File: src/charm.py [hunk 9]
```diff
@@ -197,6 +219,54 @@ def _network_attachment_definitions_from_config(self) -> list[NetworkAttachmentD
...
+        self._set_config_in_workload(upf_ip_address)
+
+    def _get_upf_ip_address_from_routing_table(self, routing_table: RoutingTable) -> Optional[str]:
+        """Returns a the UPF IP address.
+
+        If the access network exists in the routing table, it returns its gateway.
+        Otherwise, it returns the config values.
+        """
+        if networks := routing_table.get(ACCESS_NETWORK_NAME, []):
+            return networks[0]["gateway"]
```
Comment (patriciareinoso) [inline src/charm.py:267]: to check [0]

Comment (patriciareinoso) [reply src/charm.py:267]: what is the IP ADDRESS?

**Comment by gruyaume** [comment]
Closing as we won't implement this

**Pull request closed without merging**